src/Microsoft.DocAsCode.Build.Engine/MarkdownProviders/JsonServiceProvider.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.Engine
+{
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Composition;
+    using System.Text;
+
+    using Microsoft.DocAsCode.Dfm;
+    using Microsoft.DocAsCode.MarkdownLite;
+    using Microsoft.DocAsCode.Plugins;
+
+    [Export("jsonServiceProvider", typeof(IMarkdownServiceProvider))]
+    public class JsonServiceProvider: IMarkdownServiceProvider
+    {
+        public IMarkdownService CreateMarkdownService(MarkdownServiceParameters parameters)
+        {
+            return new JsonRenderService(
+                parameters.BasePath,
+                parameters.Tokens,
+                MarkdownTokenTreeValidatorFactory.Combine(TokenTreeValidator));
+        }
+
+        [ImportMany]
+        public IEnumerable<IMarkdownTokenTreeValidator> TokenTreeValidator { get; set; }
+
+        private sealed class JsonRenderService : IMarkdownService
+        {
+            private readonly DfmEngineBuilder _builder;
+
+            private readonly ImmutableDictionary<string, string> _tokens;
+
+            public JsonRenderService(string baseDir, ImmutableDictionary<string, string> tokens, IMarkdownTokenTreeValidator tokenTreeValidator)
+            {
+                _builder = DocfxFlavoredMarked.CreateBuilder(baseDir);
+                _builder.TokenTreeValidator = tokenTreeValidator;
+                _tokens = tokens;
+            }
+
+            public MarkupResult Markup(string src, string path)
+            {
+                var dependency = new HashSet<string>();
+                var json = new StringBuilder(_builder.CreateDfmEngine(new JsonRenderer() { Tokens = _tokens }).Markup(src, path, dependency));
+                if (json.Length != 0)
+                {
+                    json.Insert(0, "{\n\"name\":\"markdown\",\n\"contents\":[");
+                    json.Remove(json.Length - 1, 1);
```

Comment:
does it have default value?[https://msdn.microsoft.com/en-us/library/system.text.stringbuilder.remove(v=vs.110).aspx](https://msdn.microsoft.com/en-us/library/system.text.stringbuilder.remove(v=vs.110).aspx)

---
In reply to: [74703946](https://github.com/dotnet/docfx/pull/602#discussion_r74703946) [](ancestors = 74703946)